### PR TITLE
Module: security to prevent different types of attacks

### PR DESCRIPTION
**Pull request by Pepelux**
#### Pre-Submission Checklist

- [x] Commit message has the format required by CONTRIBUTING guide
- [x] Commits are split per component (core, individual modules, libs, utils, ...)
- [x] Each component has a single commit (if not, squash them into one commit)
- [x] No commits to README files for modules (changes must be done to docbook files
in `doc/` subfolder, the README file is autogenerated)

#### Type Of Change
- [ ] Small bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)

#### Checklist:

- [x] PR should be backported to stable branches
- [x] Tested changes locally
- [ ] Related to issue #XXXX (replace XXXX with an open issue number)

#### Description

This module has been designed to offer an additional layer of security over our communications. To achieve this, the following features are available: Blacklist to filter user agents, IP addresses, countries, domains and users. Whitelist to filter user agents, IP addresses, countries, domains and users. Blacklist of destinations where calling is not allowed. SQL injection attacks prevention.

### CODE REVIEW

**Comment by miconda** [comment]
Thanks for this submission!

Few remarks after a very brief look at the patch:

  * global variables must be declared in a .c file and the .h file has to list them with `external`, otherwise they get defined in each .c file where the .h file is included
  * it is recommended to use a prefix for all global variables and functions (which are not declared static) in order to avoid symbol conflicts. It can be the module name or a short form of it (e.g., `sec`)

One thing that is not something critical, but maybe worth discussing it before merging: is the name of the module `security` too generic for what the modules does? It does mainly filters, so eventually can be named `secfilter` (from security filter) in order to suggest better its purpose... I would like to avoid very generic names that can induce the perception it does everything expected in the scope of that name.

**Comment by Pepelux** [comment]
Thanks for the quick response

- Module has been renamed to secfilter
- Global variables are now declared in the .c files
- All non static variables now have _sec__ prefix

**Comment by henningw** [comment]
Thank you for the contribution. I will also do a quick review of the code today.

File: src/modules/secfilter/secfilter_hdr.c [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+#include "../../core/parser/parse_uri.h"
+#include "../../core/parser/parse_param.h"
+#include "../../core/parser/parse_from.h"
+#include "../../core/parser/parse_to.h"
+#include "../../core/parser/contact/parse_contact.h"
+#include "../../core/sr_module.h"
+#include "secfilter.h"
+
+
+int count_chars(char *val, char c);
```
Comment (henningw) [inline src/modules/secfilter/secfilter_hdr.c]: forward definition not needed here

File: src/modules/secfilter/secfilter_hdr.c [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+			}
+	        }
+	}
+
+	if (val) pkg_free(val);
+	return 1;
+}
+
+/* Search for illegal characters in From header */
+int check_sqli_from(struct sip_msg *msg)
```
Comment (henningw) [inline src/modules/secfilter/secfilter_hdr.c]: unify to one function, that uses $fn, $fd PVs in the kamailio configuration

File: src/modules/secfilter/secfilter_hdr.c [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+	        }
+	}
+
+	if (val) pkg_free(val);
+	return 1;
+}
+
+
+/* Search for illegal characters in Contact header */
+int check_sqli_contact(struct sip_msg *msg)
```
Comment (henningw) [inline src/modules/secfilter/secfilter_hdr.c]: See above, could be done with PVs and transformations: $ct{uri.domain}, $ct{uri.user}

File: src/modules/secfilter/secfilter_hdr.c [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+    {     
+        if (val[i] == c) cont++;
+    }
+
+    return cont;
+}
+
+
+/* Search for illegal characters in User-agent header */
+int check_sqli_ua(struct sip_msg *msg)
```
Comment (henningw) [inline src/modules/secfilter/secfilter_hdr.c]: See below, PV $ua

File: src/modules/secfilter/secfilter_hdr.c [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+		return 0;
+	}
+
+	if (val) pkg_free(val);
+	return 1;
+}
+
+
+/* Search for illegal characters in To header */
+int check_sqli_to(struct sip_msg *msg)
```
Comment (henningw) [inline src/modules/secfilter/secfilter_hdr.c]: See below, there is a PV also for To header $tu

File: src/modules/secfilter/security.c [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+        {"action_col",      PARAM_STR, &action_col },
+        {"type_col",        PARAM_STR, &type_col },
+        {"data_col",        PARAM_STR, &data_col },
+        {"dst_exact_match", PARAM_INT, &dst_exact_match },
+        {0, 0, 0}
+};
+
+/* Module exports definition */
+struct module_exports exports={
+	"security",		/* module name */
```
Comment (henningw) [inline src/modules/secfilter/security.c]: Old name

File: src/modules/secfilter/security.c [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+	}
+}
+
+
+/* Module init function */
+static int mod_init(void)
+{
+	int i;
+
+	LM_INFO("SECURITY module init\n");
```
Comment (henningw) [inline src/modules/secfilter/security.c]: Name

File: src/modules/secfilter/security.c [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+                return 0; /* do nothing for the main process */
+
+	return 1;
+}
+
+
+/* Module destroy function */
+static void mod_destroy(void)
+{
+	LM_INFO("SECURITY module destroy\n");
```
Comment (henningw) [inline src/modules/secfilter/security.c]: Name

File: src/modules/secfilter/security_rpc.c [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+
+/* Add blacklist values to database */
+void rpc_add_bl(rpc_t *rpc, void *ctx)
+{
+	char *type;
+	char *value;
+
+        if(rpc->scan(ctx, "ss", (char*)(&type), (char*)(&value))<2)
+        {
+	        rpc->fault(ctx, 0, "Invalid Parameters. Usage: security.add_bl type value\n     Example: security.add_bl user sipvicious");
```
Comment (henningw) [inline src/modules/secfilter/security_rpc.c]: Name

File: src/modules/secfilter/security_rpc.c [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+
+/* Add whitelist values to database */
+void rpc_add_wl(rpc_t *rpc, void *ctx)
+{
+	char *type;
+	char *value;
+
+        if(rpc->scan(ctx, "ss", (char*)(&type), (char*)(&value))<2)
+        {
+	        rpc->fault(ctx, 0, "Invalid Parameters. Usage: security.add_wl type value\n     Example: security.add_wl user trusted_user");
```
Comment (henningw) [inline src/modules/secfilter/security_rpc.c]: Name

File: src/modules/secfilter/secfilter.h [hunk 11]
```diff
@@ -0,0 +1,64 @@
+#ifndef _SECURITY_H
```
Comment (henningw) [inline src/modules/secfilter/secfilter.h]: Name

File: src/modules/secfilter/security.c [hunk 12]
```diff
@@ -0,0 +1,376 @@
...
+static int w_check_sqli(struct sip_msg* msg);
+
+/* Database variables */
+db_func_t db_funcs;       /* Database API functions */
+db1_con_t* db_handle=0;   /* Database connection handle */
+
+/* Exported module parameters - default values */
+int dst_exact_match = 1;
+str db_url = {NULL, 0};
+str table_name = str_init("security");
```
Comment (henningw) [inline src/modules/secfilter/security.c]: table name

File: src/modules/secfilter/security.c [hunk 13]
```diff
@@ -0,0 +1,376 @@
...
+	"Print values from database", NULL};
+
+static const char *rpc_add_bl_doc[2] = {
+	"Add new values to blacklist", NULL};
+
+static const char *rpc_add_wl_doc[2] = {
+	"Add new values to whitelist", NULL};
+
+rpc_export_t security_rpc[] = {
+        { "security.reload",  rpc_reload, rpc_reload_doc, 0},
```
Comment (henningw) [inline src/modules/secfilter/security.c]: Command names

File: src/modules/secfilter/security_hdr.c [hunk 14]
```diff
@@ -0,0 +1,294 @@
...
+#include "../../core/parser/contact/parse_contact.h"
+#include "../../core/sr_module.h"
+#include "security.h"
+
+
+int count_chars(char *val, char c);
+
+
+/* Count chars of a string */
+int count_chars(char *val, char c)
```
Comment (henningw) [inline src/modules/secfilter/security_hdr.c]: add a prefix here as well, there is already a count_chars() in another module

File: src/modules/secfilter/security_hdr.c [hunk 15]
```diff
@@ -0,0 +1,294 @@
...
+#include "../../core/parser/parse_uri.h"
+#include "../../core/parser/parse_param.h"
+#include "../../core/parser/parse_from.h"
+#include "../../core/parser/parse_to.h"
+#include "../../core/parser/contact/parse_contact.h"
+#include "../../core/sr_module.h"
+#include "security.h"
+
+
+int count_chars(char *val, char c);
```
Comment (henningw) [inline src/modules/secfilter/security_hdr.c]: forward definition not needed here

File: src/modules/secfilter/secfilter_hdr.c [hunk 16]
```diff
@@ -0,0 +1,294 @@
...
+#include "../../core/parser/contact/parse_contact.h"
+#include "../../core/sr_module.h"
+#include "secfilter.h"
+
+
+int count_chars(char *val, char c);
+
+
+/* Count chars of a string */
+int count_chars(char *val, char c)
```
Comment (henningw) [inline src/modules/secfilter/secfilter_hdr.c]: redundant implementation, move this to a header file

File: src/modules/secfilter/secfilter_db.c [hunk 17]
```diff
@@ -0,0 +1,408 @@
...
+
+	/* Add values to array */
+	for (i=0; i<rows; i++)
+	{
+		/* Blacklist */
+		if (action == 0)
+		{
+			if (!strcmp(ctype, "ua"))
+			{
+			        sec_bl_ua_list[*sec_nblUa]=strtok(strdup((char*)RES_ROWS(db_res)[i].values[0].val.string_val), "\n");
```
Comment (henningw) [inline src/modules/secfilter/secfilter_db.c]: strdup works with malloc, we need here shm_malloc as well

File: src/modules/secfilter/secfilter_db.c [hunk 18]
```diff
@@ -0,0 +1,408 @@
...
+		{
+			if (!strcmp(ctype, "ua"))
+			{
+			        sec_bl_ua_list[*sec_nblUa]=strtok(strdup((char*)RES_ROWS(db_res)[i].values[0].val.string_val), "\n"); 
+			        uppercase(sec_bl_ua_list[*sec_nblUa]);
+		        	*sec_nblUa = *sec_nblUa + 1;
+		        }
+		        if (!strcmp(ctype, "country"))
+			{
+			        sec_bl_country_list[*sec_nblCountry]=strtok(strdup((char*)RES_ROWS(db_res)[i].values[0].val.string_val), "\n");
```
Comment (henningw) [inline src/modules/secfilter/secfilter_db.c]: See above, for all strupd() calls

**Comment by Pepelux** [comment]
Thank you very much for the review. I am working on solving each point.

**Comment by miconda** [comment]
I would also like to suggest to replace that `1024` in the code with a define, it makes it easier to change if proves to be a small number.

Then the functions exported to the config should be also prefixed with `sec_` or `secf_` (e.g., secf_check_ua(...)) -- in the past few years, the functions exported by modules to config use a common prefix to be suggestive where they belong to.

**Comment by Pepelux** [comment]
All code have beed reprogrammed from scratch. Now:

- No limits to store data into blacklist or whitelist
- All module functions starts with secf_
- Change of data type and now strlist is used instead of char **
- It is possible to check SQLi in several headers and it is also possible to check only a value
- For functions that check headers, they are obtained from the structure sip_msg

Thanks a lot to @linuxmaniac for the help

**Comment by miconda** [comment]
I haven't gone into deep analysis, but the new code looks clean/well indented, exported functions have prefix, ..., so from my point of view, I am happy to merge it. Bugs can be discovered and fixed later, usually they pop up quicker when using it rather than reading code. If someone else wants to read more the code, can do it.

Anyhow, I would like to merge it manually (keeping you as the author), not from github UI (not even squashing), to have single commits for the latest version and per component. Once merged, you will get commit access so fixes and enhancements can be pushed directly by you.

**Comment by henningw** [comment]
@Pepelux Thank you for the new version, this looks really a lot better! No objections from my side for merge into the git master. @miconda - let me know if I should take care of this.

**Comment by miconda** [comment]
@henningw - thanks for your help here, as I said before, I wanted to merge it manually to split it in commits per component.

@Pepelux - I just merged the module, you should get soon an invitation on github account to join the dev team. You will be able to push commits directly, you can still use pull requests if you want others to have a look before merging. Pull requests are still recommended if you have commits for other components than secfilter module. Thanks for the contribution!

**Comment by Pepelux** [comment]
thanks all

**Pull request closed without merging**